x/valset-pref/keeper.go

```diff
@@ -89,24 +94,45 @@ func (k Keeper) GetValSetPreferencesWithDelegations(ctx sdk.Context, delegator s
 
 	// when existing delegation exists, have it based upon the existing delegation
 	// regardless of the delegator having valset pref or not
-	return types.ValidatorSetPreferences{Preferences: formatToValPrefArr(existingDelegations)}, nil
+	preferences, err := k.formatToValPrefArr(ctx, existingDelegations)
+	if err != nil {
+		return types.ValidatorSetPreferences{}, err
+	}
+	return types.ValidatorSetPreferences{Preferences: preferences}, nil
 }
 
 // formatToValPrefArr iterates over given delegations array, formats it into ValidatorPreference array.
 // Used to calculate weights for the each delegation towards validator.
 // CONTRACT: This method assumes no duplicated ValOperAddress exists in the given delegation.
-func formatToValPrefArr(delegations []stakingtypes.Delegation) []types.ValidatorPreference {
-	totalShares := sdk.NewDec(0)
+func (k Keeper) formatToValPrefArr(ctx sdk.Context, delegations []stakingtypes.Delegation) ([]types.ValidatorPreference, error) {
+	totalTokens := osmomath.NewDec(0)
+
+	// We cache token amounts for each delegation to avoid a second set of reads
+	tokenDelegations := make(map[stakingtypes.Delegation]osmomath.Dec)
 	for _, existingDelegation := range delegations {
-		totalShares = totalShares.Add(existingDelegation.Shares)
+		// Fetch validator corresponding to current delegation
+		validator, found := k.stakingKeeper.GetValidator(ctx, existingDelegation.GetValidatorAddr())
+		if !found {
+			return []types.ValidatorPreference{}, types.ValidatorNotFoundError{ValidatorAddr: existingDelegation.ValidatorAddress}
+		}
+
+		// Convert shares to underlying token amounts
+		currentDelegationTokens := validator.TokensFromSharesRoundUp(existingDelegation.Shares)
```

Comment:
I think we should be using the tokensFromShares method and track the exact Dec instead of rounding one way or the other. We are just using this to determine the ratios, so might as well be precise with it.

Reply:
I had considered this and simply default rounded up to be safe (since we're dividing by this) but it should be fine to do bankers here.

b954aaae9f2d5af5ddab238cecb2fa01175c4968